---
title: "Chapter 8 examples"
author: "Brooke Anderson"
date: "4/7/2020"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load some general packages you'll need:

```{r message = FALSE, error = FALSE}
library(tidyverse)
```


## Pasilla dataset

Install the `pasilla` package and find the file address of the example data file you'll be using. The 
`system.file` function looks for a file that was included with a package. If you want to share data using 
something other than an R binary data file, this is how the data is included in the package. The output
(`fn`) will be the filepath to the file "pasilla_gene_counts.tsv", which you downloaded as part of the
`pasilla` package.

```{r}
# Uncomment and run the following line if you need the `pasilla` package
# BiocManager::install(pkgs = "pasilla")
fn <- system.file("extdata", "pasilla_gene_counts.tsv", 
                  package = "pasilla", mustWork = TRUE)
fn
```

Now that you know where the file is, you need to read it in:

```{r message = FALSE}
counts <- fn %>% 
  read_tsv()
counts
```

Again, this data is "transposed" compared to what we'll want for a lot of statistical modeling functions. 
Each column is an observation, while each row is the measures for a specific gene (which one is given 
in the first column, `gene_id`).

There is a vignette for the `pasilla` package at: https://bioconductor.org/packages/release/data/experiment/vignettes/pasilla/inst/doc/create_objects.html
The paper that the data originally came from is available [here](https://genome.cshlp.org/content/21/2/193).

You can explore the data a bit:

```{r}
# See how many rows and columns the data has
counts %>% 
  dim()

# See a summary of each column
counts %>% 
  summary()

# See a random sample of 5 rows
counts %>% 
  sample_n(size = 5)
  

# Check out the unique letters in the gene names 
# (i.e., once you take out all the digits at the end)
counts %>% 
  pull(gene_id) %>% # Extract just the `gene_id` column as a vector
  str_remove("[0-9].+") %>% # Remove the first digit and anything after 
                            # using regular expressions 
  unique() # Look at just the unique values
```

It looks like every gene id starts with "FBgn" and then some 7-digit numeric ID.

Next, it might be helpful to sample a few of the genes
and then visualize how the values recorded for each gene vary across the samples, 
showing treated / untreated with color: 

```{r}
counts %>% 
  sample_n(size = 20) %>% 
  pivot_longer(cols = -gene_id, names_to = "sample_id", values_to = "value") %>% 
  mutate(treatment = str_remove(sample_id, "[0-9]")) %>% # Extract untreated / treated
                                                         # from sample IDs by getting
                                                         # rid of the number in each
  ggplot(aes(x = treatment, y = value)) + 
  labs(x = "", y = "") + 
  geom_point(alpha = 0.5, size = 1) + # Use some transparency (alpha) to see if points
                                      # are plotted on top of each other.
  facet_wrap(~ gene_id, scales = "free_y") + # Since scales are very different across 
                                             # genes, set scales as free on y-axis
  # Rotate the x-axis labels a bit (otherwise, they overlap)
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

Check out how much the range in the measured levels for genes varies across 
genes:

```{r message = FALSE, fig.align = "center", out.width = "4in", out.height = "3in"}
library("scales")
counts %>% 
  # Rearrange the data shape to making grouping and summarizing easier
  pivot_longer(-gene_id, names_to = "sample_id", values_to = "value") %>% 
  # Group by gene and then calculate the minimum and maximum levels of each gene
  # and the difference between those (range)
  group_by(gene_id) %>% 
  summarize(min_value = min(value),
            max_value = max(value), 
            range = max_value - min_value) %>% 
  # Plot a histogram of these ranges
  ggplot(aes(x = range)) + 
  geom_histogram(bins = 100) + 
  labs(x = "Range in read counts across samples",
       y = "Number of genes with that range\nin read counts across samples") + 
  # Use a transformation on the x-axis---otherwise, with a few outliers, it's 
  # hard to see variation among the low values. You can't just use "log" in this
  # case because you have a lot of zero values.
  scale_x_continuous(trans = "pseudo_log", labels = comma) + 
  scale_y_continuous(labels = comma) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1))




select(-mean) %>%
  unnest(data) %>%
ggplot(aes(x= sample_id, y = gene_id, fill = value)) +
  geom_tile() +
  scale_fill_viridis()
```

Read in the meta-data for these data. The meta-data are available in a different
file in the `padilla` package ("pasilla_sample_annotation.csv").

```{r message = FALSE}
annotationFile <- system.file("extdata", "pasilla_sample_annotation.csv",
                              package = "pasilla", mustWork = TRUE)
pasillaSampleAnno <- annotationFile %>% 
  read_csv()
pasillaSampleAnno
```

This file gives some general information about each sample. This includes the file name
where the sample's data was saved (probably from the sampling equipment), 
the treatment or condition (treated / untreated), the type of read (single-read
or paired-end), the number of lanes, total number of reads, and exon counts. 

It can be hard to work with column names with spaces in the middle, so I 
recommend renaming the columns (replacing every space with an underscore will 
work):

```{r}
# Current column names (they have spaces)
pasillaSampleAnno %>% 
  colnames()

# Rename columns to get rid of spaces
pasillaSampleAnno <- pasillaSampleAnno %>% 
  # \\s stands for a space in R regular expressions
  rename_all(.funs = str_replace_all, "\\s", "_") 

# New column names (no more spaces)
pasillaSampleAnno %>% 
  colnames()
```

We can see some summaries of the data by condition (treated / untreated) and 
type (single-read / pair-ended):

```{r}
pasillaSampleAnno %>% 
  # Group the data and use `count` to count the number of rows with each combo
  group_by(condition, type) %>% 
  count() %>% 
  ungroup() %>% 
  # Reshape the data to make a prettier table
  pivot_wider(names_from = type, values_from = n) %>% 
  # Use `kable` from the `knitr` package to output as a pretty table
  knitr::kable()
```

Create a `DESeqDataSet` object with both the measurements (from `counts`) and 
the meta-data (from `pasillaSampleAnno`). This *data container* is a special 
type of object class that will help keep everything in order as you move through 
the analysis (and helps with managing the size of the data). 

```{r message = FALSE}
library("DESeq2")

# Need to remember what the column names of `counts` are (so we can 
# use the same ones for the meta-data)
colnames(counts)

# Clean up the meta-data a bit before we add it to this special object class
pasillaSampleAnno <- pasillaSampleAnno %>% 
  rename(file = "sample_id") %>% 
  # Use regular expressions to clean up sample_id name (so it matches
  # the column names in `counts`)
  mutate(sample_id = str_remove(sample_id, "fb")) %>% 
  # Evidently, the "-"s in the `type` column may cause problems in DESeq, 
  # so shorten those labels
  mutate(type = str_remove(type, "-.+")) %>% 
  # Convert the `condition` and `type` columns to factors and 
  # set the level order by hand 
  mutate(condition = condition %>% 
           as_factor() %>% 
           fct_relevel("untreated", "treated"),
         type = type %>% 
           as_factor() %>% 
           fct_relevel("single", "paired")) %>% 
  # Finally, you need to make sure the rows of this dataframe
  # are in the same order as the columns in the read count data matrix.
  # One way to do this is by making the column names of the count data
  # matrix (without the gene_id column) into a dataframe, with numbers
  # giving the order, join that with this data, then reorder by that
  # order number
  full_join(counts %>% 
              # Get the column names from `counts`
              colnames() %>% 
              # Get rid of the first column name (which is the `gene_id` column)
              `[`(-1) %>% 
              # `enframe` will convert to a dataframe (from a vector)
              enframe(name = "order"), 
            by = c("sample_id" = "value")) %>% 
  # Rearrange by order, then get rid of that column
  arrange(order) %>% 
  select(-order)

# Here's what the meta-data looks like now:
pasillaSampleAnno

# Put all the required components into a DESeqDataSet object so you can 
# run DESeq on it
pasilla <- counts %>% 
  # For the read data, it's all in counts, but we need to take off the 
  # first column (with the gene IDs) and then convert to a matrix to 
  # input it into this object class
  column_to_rownames("gene_id") %>% 
  as.matrix() %>% 
  # Put everything into a DESeqDataSet object
  DESeqDataSetFromMatrix(colData = pasillaSampleAnno, 
                         design = ~ condition)

class(pasilla)
```


## DESeq

Apply `DESeq` algorithm to the data: 

```{r}
pasilla <- pasilla %>% 
  DESeq()
```

You can use *extractor functions* like `results` to pull out specific data from 
the resulting object. For example, you can pull out gene-specific estimates of
differences between treated and untreated samples with:

```{r}
pasilla %>% 
  results()
```

If you want to work more easily with this output, you can convert it to a 
tibble with `as_tibble`:

```{r}
pasilla %>% 
  results() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene_id") %>% 
  as_tibble()
```

Check out the top few genes in terms of adjusted p-values for the Wald test comparing
the untreated versus treated samples. Since smaller p-values are more interesting, 
pick out the smallest:

```{r}
pasilla %>% 
  results() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene_id") %>% 
  filter(!is.na(padj)) %>% 
  as_tibble() %>% 
  arrange(padj) %>% 
  dplyr::slice(1:5)
```

Histogram of p-values: 

```{r fig.width = 3.5, fig.height = 2, fig.align = "center", message = FALSE, warning = FALSE}
pasilla %>% 
  results() %>% 
  as_tibble() %>% 
  ggplot(aes(x = pvalue)) + 
  geom_histogram(binwidth = 0.01, fill = "Royalblue", boundary = 0)
```

MA plot (but using ggplot instead of `plotMA` from `DESeq2` package---the x-axis transform might not be identical...):

```{r warning = FALSE}
pasilla %>% 
  results() %>% 
  as_tibble() %>% 
  mutate(low_padj = padj < 0.1) %>% 
  mutate(log2FoldChange = ifelse(log2FoldChange > 2, 2, log2FoldChange),
         log2FoldChange = ifelse(log2FoldChange < -2, -2, log2FoldChange)) %>% 
  ggplot(aes(x = baseMean, y = log2FoldChange, color = low_padj)) + 
  geom_point(size = 0.5) + 
  ylim(c(-2, 2)) + 
  scale_color_manual(values = c("black", "red")) + 
  scale_x_continuous(trans = "pseudo_log") + 
  geom_hline(yintercept = 0, color = "red", size = 2, alpha = 0.4)
```


